cogfeeds/cogfeeds.py

```diff
@@ -0,0 +1,520 @@
+
+
+class EmbedBuilder(object):
+    """
+    Builds the embed image for the provided and relevant data.
+    The YAML will have been validated before it reaches this point.
+    Irrelvant keys are ignored as we are only getting the keys we need
+    through dict.get()
+
+    Attributes:
+    data: dict
+        The extracted data from the yaml conversion
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.data = kwargs.get("data")
+        self.name = self.data.get("name")
+        self.author = self.data.get("author")
+        self.description = self.data.get("description")
+        self.requirements = self.data.get("requirements")
+        self.tags = self.data.get("tags")
+        self.end_user_data_statement = self.data.get("end_user_data_statement")
+        self.install_guide = self.data.get("install_guide", False)
+        self.github_url = self.data.get("github_url")
+        self.smart_link_title = self.data.get("smart_link_title")
+        # list of all viable keys
+        self.key_list = [
+            "name", 
+            "author",
+            "description",
+            "requirements",
+            "tags",
+            "end_user_data_statement",
+            "install_guide",
+            "github_url",
+            "smart_link_title",
+        ]
+        # Removal build
+        self.reason = self.data.get("reason")
+        self.additional_info = self.data.get("additional_info")
+        # Misc
+        self.publish_type = kwargs.get("publish_type")
+
+    def __repr__(self) -> str:
+        return f"<class {self.__class__.__name__}(data={self.data})>"
+
+    @property
+    def __str__(self):
+        return self.__repr__
+
+    def __len__(self) -> int:
+        return len(self.data.keys())
+
+    def format_build(self, ctx):
+        if not self.name:
+            raise ParserRequiredKeyError("name")
+        else:
+            if not isinstance(self.name, str):
+                raise ParserInvalidTypeError(
+                    field="name",
+                    invalid_type=type(self.name),
+                    supported_types=(str,)
+                )
+
+        if self.author:
+            if not isinstance(self.author, (str, list)):
+                raise ParserInvalidTypeError(
+                    field="author",
+                    invalid_type=type(self.author),
+                    supported_types=(str, list)
+                )
+            if isinstance(self.author, list):
+                for author in self.author:
+                    if not isinstance(author, str):
+                        raise ParserInvalidItemError(
+                            item=author,
+                            array="author",
+                            literal_array=self.author
+                        )
+        else:
+            self.author = ctx.author.name
+
+        if self.publish_type == "removal":
+            if self.reason:
+                if not isinstance(self.reason, str):
```

Comment:
This could be changed to `if self.reason and not isinstance(self.reason, str):`, however, this isn't necessary